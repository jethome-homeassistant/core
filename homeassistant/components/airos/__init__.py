"""The Ubiquiti airOS integration."""

from __future__ import annotations

import logging

from airos.airos8 import AirOS8

from homeassistant.const import (
    CONF_HOST,
    CONF_PASSWORD,
    CONF_SSL,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
    Platform,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import device_registry as dr, entity_registry as er
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DEFAULT_SSL, DEFAULT_VERIFY_SSL, DOMAIN, SECTION_ADVANCED_SETTINGS
from .coordinator import AirOSConfigEntry, AirOSDataUpdateCoordinator

_PLATFORMS: list[Platform] = [
    Platform.BINARY_SENSOR,
    Platform.SENSOR,
]

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: AirOSConfigEntry) -> bool:
    """Set up Ubiquiti airOS from a config entry."""

    # By default airOS 8 comes with self-signed SSL certificates,
    # with no option in the web UI to change or upload a custom certificate.
    session = async_get_clientsession(
        hass, verify_ssl=entry.data[SECTION_ADVANCED_SETTINGS][CONF_VERIFY_SSL]
    )

    airos_device = AirOS8(
        host=entry.data[CONF_HOST],
        username=entry.data[CONF_USERNAME],
        password=entry.data[CONF_PASSWORD],
        session=session,
        use_ssl=entry.data[SECTION_ADVANCED_SETTINGS][CONF_SSL],
    )

    coordinator = AirOSDataUpdateCoordinator(hass, entry, airos_device)
    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, _PLATFORMS)

    return True


async def async_migrate_entry(hass: HomeAssistant, entry: AirOSConfigEntry) -> bool:
    """Migrate old config entry."""

    # This means the user has downgraded from a future version
    if entry.version > 2:
        return False

    # 1.1 Migrate config_entry to add advanced ssl settings
    if entry.version == 1 and entry.minor_version == 1:
        new_minor_version = 2
        new_data = {**entry.data}
        advanced_data = {
            CONF_SSL: DEFAULT_SSL,
            CONF_VERIFY_SSL: DEFAULT_VERIFY_SSL,
        }
        new_data[SECTION_ADVANCED_SETTINGS] = advanced_data

        hass.config_entries.async_update_entry(
            entry,
            data=new_data,
            minor_version=new_minor_version,
        )

    # 2.1 Migrate binary_sensor entity unique_id from device_id to mac_address
    #     Step 1 - migrate binary_sensor entity unique_id
    #     Step 2 - migrate device entity identifier
    if entry.version == 1:
        new_version = 2
        new_minor_version = 1

        mac_adress = dr.format_mac(entry.unique_id)

        device_registry = dr.async_get(hass)
        if device_entry := device_registry.async_get_device(
            connections={(dr.CONNECTION_NETWORK_MAC, mac_adress)}
        ):
            old_device_id = next(
                (
                    device_id
                    for domain, device_id in device_entry.identifiers
                    if domain == DOMAIN
                ),
            )

            @callback
            def update_unique_id(
                entity_entry: er.RegistryEntry,
            ) -> dict[str, str] | None:
                """Update unique id from device_id to mac address."""
                if old_device_id and entity_entry.unique_id.startswith(old_device_id):
                    suffix = entity_entry.unique_id.removeprefix(old_device_id)
                    new_unique_id = f"{mac_adress}{suffix}"
                    return {"new_unique_id": new_unique_id}
                return None

            await er.async_migrate_entries(hass, entry.entry_id, update_unique_id)

            new_identifiers = device_entry.identifiers.copy()
            new_identifiers.discard((DOMAIN, old_device_id))
            new_identifiers.add((DOMAIN, mac_adress))
            device_registry.async_update_device(
                device_entry.id, new_identifiers=new_identifiers
            )

        hass.config_entries.async_update_entry(
            entry, version=new_version, minor_version=new_minor_version
        )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: AirOSConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, _PLATFORMS)
