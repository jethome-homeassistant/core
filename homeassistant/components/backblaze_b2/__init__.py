"""The Backblaze B2 integration."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Any

from b2sdk.v2 import B2Api, Bucket, InMemoryAccountInfo, exception

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers.event import async_track_time_interval

from .const import (
    BACKBLAZE_REALM,
    CONF_APPLICATION_KEY,
    CONF_BUCKET,
    CONF_KEY_ID,
    DATA_BACKUP_AGENT_LISTENERS,
    DOMAIN,
)
from .repairs import (
    async_check_for_repair_issues,
    create_bucket_access_restricted_issue,
    create_bucket_not_found_issue,
)

_LOGGER = logging.getLogger(__name__)

type BackblazeConfigEntry = ConfigEntry[Bucket]


async def async_setup_entry(hass: HomeAssistant, entry: BackblazeConfigEntry) -> bool:
    """Set up Backblaze B2 from a config entry."""

    info = InMemoryAccountInfo()
    b2_api = B2Api(info)

    def _authorize_and_get_bucket_sync() -> Bucket:
        """Synchronously authorize the Backblaze B2 account and retrieve the bucket.

        This function runs in the event loop's executor as b2sdk operations are blocking.
        """
        b2_api.authorize_account(
            BACKBLAZE_REALM,
            entry.data[CONF_KEY_ID],
            entry.data[CONF_APPLICATION_KEY],
        )
        return b2_api.get_bucket_by_name(entry.data[CONF_BUCKET])

    try:
        bucket = await hass.async_add_executor_job(_authorize_and_get_bucket_sync)
    except exception.Unauthorized as err:
        raise ConfigEntryAuthFailed(
            translation_domain=DOMAIN,
            translation_key="invalid_credentials",
        ) from err
    except exception.RestrictedBucket as err:
        create_bucket_access_restricted_issue(hass, entry, err.bucket_name)
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="restricted_bucket",
            translation_placeholders={
                "restricted_bucket_name": err.bucket_name,
            },
        ) from err
    except exception.NonExistentBucket as err:
        create_bucket_not_found_issue(hass, entry, entry.data[CONF_BUCKET])
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="invalid_bucket_name",
        ) from err
    except exception.ConnectionReset as err:
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="cannot_connect",
        ) from err
    except exception.MissingAccountData as err:
        raise ConfigEntryAuthFailed(
            translation_domain=DOMAIN,
            translation_key="invalid_auth",
        ) from err

    entry.runtime_data = bucket

    def _async_notify_backup_listeners() -> None:
        """Notify any registered backup agent listeners."""
        _LOGGER.debug("Notifying backup listeners for entry %s", entry.entry_id)
        for listener in hass.data.get(DATA_BACKUP_AGENT_LISTENERS, []):
            listener()

    entry.async_on_unload(entry.async_on_state_change(_async_notify_backup_listeners))

    async def _periodic_issue_check(_now: Any) -> None:
        """Periodically check for repair issues."""
        await async_check_for_repair_issues(hass, entry)

    entry.async_on_unload(
        async_track_time_interval(hass, _periodic_issue_check, timedelta(minutes=30))
    )

    hass.async_create_task(async_check_for_repair_issues(hass, entry))

    return True


async def async_unload_entry(hass: HomeAssistant, entry: BackblazeConfigEntry) -> bool:
    """Unload a Backblaze B2 config entry.

    Any resources directly managed by this entry that need explicit shutdown
    would be handled here. In this case, the `async_on_state_change` listener
    handles the notification logic on unload.
    """
    return True
