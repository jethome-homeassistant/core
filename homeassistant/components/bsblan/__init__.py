"""The BSB-Lan integration."""

import dataclasses

from bsblan import (
    BSBLAN,
    BSBLANAuthError,
    BSBLANConfig,
    BSBLANConnectionError,
    BSBLANError,
    Device,
    Info,
    StaticState,
)

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_HOST,
    CONF_PASSWORD,
    CONF_PORT,
    CONF_USERNAME,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import (
    ConfigEntryAuthFailed,
    ConfigEntryError,
    ConfigEntryNotReady,
)
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import CONF_PASSKEY, DOMAIN
from .coordinator import BSBLanFastCoordinator, BSBLanSlowCoordinator

PLATFORMS = [Platform.CLIMATE, Platform.SENSOR, Platform.WATER_HEATER]

type BSBLanConfigEntry = ConfigEntry[BSBLanData]


@dataclasses.dataclass
class BSBLanData:
    """BSBLan data stored in the Home Assistant data object."""

    fast_coordinator: BSBLanFastCoordinator
    slow_coordinator: BSBLanSlowCoordinator
    client: BSBLAN
    device: Device
    info: Info
    static: StaticState


async def async_setup_entry(hass: HomeAssistant, entry: BSBLanConfigEntry) -> bool:
    """Set up BSB-Lan from a config entry."""

    # create config using BSBLANConfig
    config = BSBLANConfig(
        host=entry.data[CONF_HOST],
        passkey=entry.data[CONF_PASSKEY],
        port=entry.data[CONF_PORT],
        username=entry.data.get(CONF_USERNAME),
        password=entry.data.get(CONF_PASSWORD),
    )

    # create BSBLAN client
    session = async_get_clientsession(hass)
    bsblan = BSBLAN(config, session)

    try:
        # Initialize the client first - this sets up internal caches and validates the connection
        await bsblan.initialize()
        # Fetch all required device metadata
        device = await bsblan.device()
        info = await bsblan.info()
        static = await bsblan.static_values()
    except BSBLANConnectionError as err:
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="setup_connection_error",
            translation_placeholders={"host": entry.data[CONF_HOST]},
        ) from err
    except BSBLANAuthError as err:
        raise ConfigEntryAuthFailed(
            translation_domain=DOMAIN,
            translation_key="setup_auth_error",
        ) from err
    except TimeoutError as err:
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="setup_connection_error",
            translation_placeholders={"host": entry.data[CONF_HOST]},
        ) from err
    except BSBLANError as err:
        raise ConfigEntryError(
            translation_domain=DOMAIN,
            translation_key="setup_general_error",
        ) from err

    # Create coordinators with the already-initialized client
    fast_coordinator = BSBLanFastCoordinator(hass, entry, bsblan)
    slow_coordinator = BSBLanSlowCoordinator(hass, entry, bsblan)

    # Perform first refresh of both coordinators
    await fast_coordinator.async_config_entry_first_refresh()

    # Try to refresh slow coordinator, but don't fail if DHW is not available
    # This allows the integration to work even if the device doesn't support DHW
    await slow_coordinator.async_refresh()

    entry.runtime_data = BSBLanData(
        client=bsblan,
        fast_coordinator=fast_coordinator,
        slow_coordinator=slow_coordinator,
        device=device,
        info=info,
        static=static,
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: BSBLanConfigEntry) -> bool:
    """Unload BSBLAN config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
