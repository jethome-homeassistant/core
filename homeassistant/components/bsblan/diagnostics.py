"""Diagnostics support for BSBLan."""

from __future__ import annotations

from typing import Any

from homeassistant.core import HomeAssistant

from . import BSBLanConfigEntry


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: BSBLanConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    data = entry.runtime_data

    # Build diagnostic data from both coordinators
    diagnostics = {
        "info": data.info.to_dict(),
        "device": data.device.to_dict(),
        "fast_coordinator_data": {
            "state": data.fast_coordinator.data.state.to_dict(),
            "sensor": data.fast_coordinator.data.sensor.to_dict(),
            "dhw": data.fast_coordinator.data.dhw.to_dict(),
        },
        "static": data.static.to_dict(),
    }

    # Add DHW config and schedule from slow coordinator if available
    if data.slow_coordinator.data:
        slow_data = {}
        if data.slow_coordinator.data.dhw_config:
            slow_data["dhw_config"] = data.slow_coordinator.data.dhw_config.to_dict()
        if data.slow_coordinator.data.dhw_schedule:
            slow_data["dhw_schedule"] = (
                data.slow_coordinator.data.dhw_schedule.to_dict()
            )
        if slow_data:
            diagnostics["slow_coordinator_data"] = slow_data

    return diagnostics
