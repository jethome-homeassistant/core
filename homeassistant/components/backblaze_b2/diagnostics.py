"""Diagnostics support for Backblaze B2."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.core import HomeAssistant

from . import BackblazeConfigEntry
from .const import CONF_APPLICATION_KEY, CONF_KEY_ID

TO_REDACT_ENTRY_DATA = {CONF_APPLICATION_KEY, CONF_KEY_ID}
TO_REDACT_ACCOUNT_DATA_ALLOWED = {"bucketId", "bucketName", "namePrefix"}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: BackblazeConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    bucket = entry.runtime_data

    try:
        bucket_info = {
            "name": bucket.name,
            "id": bucket.id_,
            "type": bucket.type_,
            "cors_rules": bucket.cors_rules,
            "lifecycle_rules": bucket.lifecycle_rules,
            "revision": bucket.revision,
        }

        account_info = bucket.api.account_info
        account_data: dict[str, Any] = {
            "account_id": account_info.get_account_id(),
            "api_url": account_info.get_api_url(),
            "download_url": account_info.get_download_url(),
            "minimum_part_size": account_info.get_minimum_part_size(),
            "allowed": account_info.get_allowed(),
        }

        if isinstance(account_data["allowed"], dict):
            account_data["allowed"] = async_redact_data(
                account_data["allowed"], TO_REDACT_ACCOUNT_DATA_ALLOWED
            )

    except (AttributeError, TypeError, ValueError, KeyError):
        bucket_info = {"name": "unknown", "id": "unknown"}
        account_data = {"error": "Failed to retrieve detailed account information"}

    return {
        "entry_data": async_redact_data(entry.data, TO_REDACT_ENTRY_DATA),
        "entry_options": entry.options,
        "bucket_info": bucket_info,
        "account_info": account_data,
    }
