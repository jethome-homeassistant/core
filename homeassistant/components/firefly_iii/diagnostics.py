"""Diagnostics for the Firefly III integration."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_API_KEY, CONF_URL
from homeassistant.core import HomeAssistant

from . import FireflyConfigEntry
from .coordinator import FireflyDataUpdateCoordinator

TO_REDACT = [CONF_API_KEY, CONF_URL]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: FireflyConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: FireflyDataUpdateCoordinator = entry.runtime_data

    return {
        "config_entry": async_redact_data(entry.as_dict(), TO_REDACT),
        "data": {"primary_currency": coordinator.data.primary_currency.to_dict()},
    }
