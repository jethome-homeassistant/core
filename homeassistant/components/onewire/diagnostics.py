"""Diagnostics support for 1-Wire."""

from __future__ import annotations

from dataclasses import asdict
from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_HOST
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr

from .onewirehub import OneWireConfigEntry

TO_REDACT = {CONF_HOST}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: OneWireConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    onewire_hub = entry.runtime_data

    return {
        "entry": {
            "title": entry.title,
            "data": async_redact_data(entry.data, TO_REDACT),
            "options": {**entry.options},
        },
        "devices": [asdict(device_details) for device_details in onewire_hub.devices],
    }


async def async_get_device_diagnostics(
    hass: HomeAssistant, entry: OneWireConfigEntry, device_entry: dr.DeviceEntry
) -> dict[str, Any]:
    """Return diagnostics for a device."""

    onewire_hub = entry.runtime_data

    return {
        "entry": {
            "title": entry.title,
            "data": async_redact_data(entry.data, TO_REDACT),
            "options": {**entry.options},
        },
        "device": asdict(
            next(
                device_details
                for device_details in onewire_hub.devices
                if device_details.id[3:] == device_entry.serial_number
            )
        ),
    }
