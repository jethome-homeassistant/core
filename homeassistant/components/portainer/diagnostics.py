"""Diagnostics for the Portainer integration."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_API_TOKEN
from homeassistant.core import HomeAssistant

from . import PortainerConfigEntry
from .coordinator import PortainerCoordinator

TO_REDACT = [CONF_API_TOKEN]


def _serialize_coordinator(coordinator: PortainerCoordinator) -> dict[str, Any]:
    """Serialize coordinator data into a JSON-safe structure."""

    serialized_endpoints: list[dict[str, Any]] = []
    for endpoint_id, endpoint_data in coordinator.data.items():
        serialized_endpoints.append(
            {
                "id": endpoint_id,
                "name": endpoint_data.name,
                "endpoint": {
                    "status": endpoint_data.endpoint.status,
                    "url": endpoint_data.endpoint.url,
                    "public_url": endpoint_data.endpoint.public_url,
                },
                "containers": [
                    {
                        "id": container.id,
                        "names": list(container.names or []),
                        "image": container.image,
                        "state": container.state,
                        "status": container.status,
                    }
                    for container in endpoint_data.containers.values()
                ],
            }
        )

    return {"endpoints": serialized_endpoints}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: PortainerConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a Portainer config entry."""

    return {
        "config_entry": async_redact_data(config_entry.as_dict(), TO_REDACT),
        "coordinator": _serialize_coordinator(config_entry.runtime_data),
    }
