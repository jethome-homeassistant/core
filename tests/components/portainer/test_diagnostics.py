"""Test the Portainer component diagnostics."""

from unittest.mock import AsyncMock

from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from homeassistant.core import HomeAssistant

from . import setup_integration

from tests.common import MockConfigEntry
from tests.components.diagnostics import get_diagnostics_for_config_entry
from tests.typing import ClientSessionGenerator


async def test_get_config_entry_diagnostics(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_portainer_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
    hass_client: ClientSessionGenerator,
) -> None:
    """Test if get_config_entry_diagnostics returns the correct data."""
    await setup_integration(hass, mock_config_entry)

    diagnostics_entry = await get_diagnostics_for_config_entry(
        hass, hass_client, mock_config_entry
    )
    assert diagnostics_entry == snapshot(
        exclude=props("created_at", "modified_at"),
    )
