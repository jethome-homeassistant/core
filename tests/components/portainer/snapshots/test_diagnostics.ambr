# serializer version: 1
# name: test_get_config_entry_diagnostics
  dict({
    'config_entry': dict({
      'data': dict({
        'api_token': '**REDACTED**',
        'url': 'https://127.0.0.1:9000/',
        'verify_ssl': True,
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'portainer',
      'entry_id': 'portainer_test_entry_123',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'Portainer test',
      'unique_id': 'test_api_token',
      'version': 2,
    }),
    'coordinator': dict({
      'endpoints': list([
        dict({
          'containers': list([
            dict({
              'id': 'aa86eacfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf',
              'image': 'docker.io/library/ubuntu:latest',
              'names': list([
                '/funny_chatelet',
              ]),
              'state': 'running',
              'status': 'Up 4 days',
            }),
            dict({
              'id': 'bb97facfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf',
              'image': 'docker.io/library/nginx:latest',
              'names': list([
                '/serene_banach',
              ]),
              'state': 'running',
              'status': 'Up 2 days',
            }),
            dict({
              'id': 'cc08facfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf',
              'image': 'docker.io/library/postgres:15',
              'names': list([
                '/stoic_turing',
              ]),
              'state': 'running',
              'status': 'Up 1 day',
            }),
            dict({
              'id': 'dd19facfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf',
              'image': 'docker.io/library/redis:7',
              'names': list([
                '/focused_einstein',
              ]),
              'state': 'running',
              'status': 'Up 12 hours',
            }),
            dict({
              'id': 'ee20facfb3b3ed4cd362c1e88fc89a53908ad05fb3a4103bca3f9b28292d14bf',
              'image': 'docker.io/library/python:3.13-slim',
              'names': list([
                '/practical_morse',
              ]),
              'state': 'running',
              'status': 'Up 6 hours',
            }),
          ]),
          'endpoint': dict({
            'public_url': 'docker.mydomain.tld:2375',
            'status': 1,
            'url': 'docker.mydomain.tld:2375',
          }),
          'id': 1,
          'name': 'my-environment',
        }),
      ]),
    }),
  })
# ---
